File: app/src/main/java/dev/ragnarok/fenrir/adapter/MenuListAdapter.java
package dev.ragnarok.fenrir.adapter;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.card.MaterialCardView;
import com.squareup.picasso3.Transformation;

import java.util.List;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter;
import dev.ragnarok.fenrir.model.drawer.AbsMenuItem;
import dev.ragnarok.fenrir.model.drawer.IconMenuItem;
import dev.ragnarok.fenrir.model.drawer.NoIconMenuItem;
import dev.ragnarok.fenrir.model.drawer.RecentChat;
import dev.ragnarok.fenrir.picasso.PicassoInstance;
import dev.ragnarok.fenrir.settings.CurrentTheme;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.Utils;

public class MenuListAdapter extends RecyclerBindableAdapter<AbsMenuItem, RecyclerView.ViewHolder> {

    private final ActionListener actionListener;
    private final int colorPrimary;
    private final int colorSurface;
    private final int colorOnPrimary;
    private final int colorOnSurface;
    private final int dp;
    private final Transformation transformation;
    private final boolean paging;
    private final boolean noStroke;

    public MenuListAdapter(@NonNull Context context, @NonNull List<AbsMenuItem> pageItems, @NonNull ActionListener actionListener, boolean paging) {
        super(pageItems);
        colorPrimary = CurrentTheme.getColorPrimary(context);
        colorSurface = CurrentTheme.getColorSurface(context);
        colorOnPrimary = CurrentTheme.getColorOnPrimary(context);
        colorOnSurface = CurrentTheme.getColorOnSurface(context);
        dp = (int) Utils.dpToPx(1, context);
        transformation = CurrentTheme.createTransformationForAvatar();
        this.actionListener = actionListener;
        this.paging = paging;
        noStroke = Settings.get().other().is_side_no_stroke();
    }

    @Override
    protected void onBindItemViewHolder(RecyclerView.ViewHolder holder, int position, int type) {
        AbsMenuItem item = getItem(position);
        holder.itemView.setSelected(item.isSelected());

        switch (type) {
            case AbsMenuItem.TYPE_ICON:
                if ((paging || !noStroke)) {
                    bindIconHolder((NormalHolder) holder, (IconMenuItem) item);
                } else {
                    bindIconNoStrokeHolder((NormalNoStrokeHolder) holder, (IconMenuItem) item);
                }
                break;
            case AbsMenuItem.TYPE_RECENT_CHAT:
                bindRecentChat((RecentChatHolder) holder, (RecentChat) item);
                break;
            case AbsMenuItem.TYPE_WITHOUT_ICON:
                bindWithoutIcon((NoIconHolder) holder, (NoIconMenuItem) item);
                break;
        }
    }

    private void bindWithoutIcon(NoIconHolder holder, NoIconMenuItem item) {
        holder.txTitle.setText(item.getTitle());
        holder.txTitle.setTextColor(item.isSelected() ? colorOnPrimary : colorOnSurface);
        holder.contentRoot.setOnClickListener(v -> actionListener.onDrawerItemClick(item));
        holder.contentRoot.setOnLongClickListener(view -> {
            actionListener.onDrawerItemLongClick(item);
            return true;
        });
    }

    private void bindIconNoStrokeHolder(NormalNoStrokeHolder holder, IconMenuItem item) {
        holder.txtTitle.setText(item.getTitle());
        holder.txtTitle.setTextColor(item.isSelected() ? colorOnPrimary : colorOnSurface);

        holder.tvCount.setVisibility(item.getCount() > 0 ? View.VISIBLE : View.GONE);
        holder.tvCount.setText(String.valueOf(item.getCount()));
        holder.tvCount.setTextColor(item.isSelected() ? colorOnPrimary : colorPrimary);

        holder.imgIcon.setImageResource(item.getIcon());
        holder.imgIcon.setColorFilter(item.isSelected() ? colorOnPrimary : colorOnSurface);

        holder.contentRoot.setBackgroundColor(item.isSelected() ? colorPrimary : colorSurface);
        holder.contentRoot.setOnClickListener(v -> actionListener.onDrawerItemClick(item));
        holder.contentRoot.setOnLongClickListener(view -> {
            actionListener.onDrawerItemLongClick(item);
            return true;
        });
    }

    private void bindIconHolder(NormalHolder holder, IconMenuItem item) {
        holder.txtTitle.setText(item.getTitle());
        holder.txtTitle.setTextColor(item.isSelected() ? colorOnPrimary : colorOnSurface);

        holder.tvCount.setVisibility(item.getCount() > 0 ? View.VISIBLE : View.GONE);
        holder.tvCount.setText(String.valueOf(item.getCount()));
        holder.tvCount.setTextColor(item.isSelected() ? colorOnPrimary : colorPrimary);

        holder.imgIcon.setImageResource(item.getIcon());
        holder.imgIcon.setColorFilter(item.isSelected() ? colorOnPrimary : colorOnSurface);

        holder.contentRoot.setCardBackgroundColor(item.isSelected() ? colorPrimary : colorSurface);
        holder.contentRoot.setStrokeWidth(item.isSelected() ? 0 : dp);
        holder.contentRoot.setOnClickListener(v -> actionListener.onDrawerItemClick(item));
        holder.contentRoot.setOnLongClickListener(view -> {
            actionListener.onDrawerItemLongClick(item);
            return true;
        });
    }

    private void bindRecentChat(RecentChatHolder holder, RecentChat item) {
        holder.tvChatTitle.setText(item.getTitle());
        holder.tvChatTitle.setTextColor(item.isSelected() ? colorOnPrimary : colorOnSurface);

        if (Utils.isEmpty(item.getIconUrl())) {
            PicassoInstance.with()
                    .load(R.drawable.ic_group_chat)
                    .transform(transformation)
                    .into(holder.ivChatImage);
        } else {
            PicassoInstance.with()
                    .load(item.getIconUrl())
                    .transform(transformation)
                    .into(holder.ivChatImage);
        }

        ((MaterialCardView) holder.contentRoot).setCardBackgroundColor(item.isSelected() ? colorPrimary : colorSurface);
        holder.contentRoot.setOnClickListener(v -> actionListener.onDrawerItemClick(item));
        holder.contentRoot.setOnLongClickListener(view -> {
            actionListener.onDrawerItemLongClick(item);
            return true;
        });
    }

    @Override
    protected RecyclerView.ViewHolder viewHolder(View view, int type) {
        switch (type) {
            case AbsMenuItem.TYPE_DIVIDER:
                return new DividerHolder(view);
            case AbsMenuItem.TYPE_RECENT_CHAT:
                return new RecentChatHolder(view);
            case AbsMenuItem.TYPE_ICON:
                return (paging || !noStroke) ? new NormalHolder(view) : new NormalNoStrokeHolder(view);
            case AbsMenuItem.TYPE_WITHOUT_ICON:
                return new NoIconHolder(view);
        }
        throw new IllegalStateException();
    }

    @Override
    protected int layoutId(int type) {
        switch (type) {
            case AbsMenuItem.TYPE_DIVIDER:
                return (noStroke ? R.layout.drawer_list_item_divider_no_stroke : R.layout.drawer_list_item_divider);
            case AbsMenuItem.TYPE_RECENT_CHAT:
                return R.layout.item_navigation_recents;
            case AbsMenuItem.TYPE_ICON:
                return paging ? R.layout.item_navigation : (noStroke ? R.layout.drawer_list_item_no_stroke : R.layout.drawer_list_item);
            case AbsMenuItem.TYPE_WITHOUT_ICON:
                return R.layout.drawer_list_item_without_icon;
        }

        throw new IllegalStateException();
    }

    @Override
    protected int getItemType(int position) {
        return getItem(position - getHeadersCount()).getType();
    }

    public interface ActionListener {
        void onDrawerItemClick(AbsMenuItem item);

        void onDrawerItemLongClick(AbsMenuItem item);
    }

    private static class NormalHolder extends RecyclerView.ViewHolder {

        final ImageView imgIcon;
        final TextView txtTitle;
        final TextView tvCount;
        final MaterialCardView contentRoot;

        NormalHolder(View view) {
            super(view);
            contentRoot = view.findViewById(R.id.content_root);
            imgIcon = view.findViewById(R.id.icon);
            txtTitle = view.findViewById(R.id.title);
            tvCount = view.findViewById(R.id.counter);
        }
    }

    private static class NormalNoStrokeHolder extends RecyclerView.ViewHolder {

        final ImageView imgIcon;
        final TextView txtTitle;
        final TextView tvCount;
        final ViewGroup contentRoot;

        NormalNoStrokeHolder(View view) {
            super(view);
            contentRoot = view.findViewById(R.id.content_root);
            imgIcon = view.findViewById(R.id.icon);
            txtTitle = view.findViewById(R.id.title);
            tvCount = view.findViewById(R.id.counter);
        }
    }

    private static class DividerHolder extends RecyclerView.ViewHolder {

        DividerHolder(View itemView) {
            super(itemView);
        }
    }

    private static class RecentChatHolder extends RecyclerView.ViewHolder {

        final TextView tvChatTitle;
        final ImageView ivChatImage;
        final View contentRoot;

        RecentChatHolder(View itemView) {
            super(itemView);
            contentRoot = itemView.findViewById(R.id.content_root);
            tvChatTitle = itemView.findViewById(R.id.title);
            ivChatImage = itemView.findViewById(R.id.avatar);
        }
    }

    private static class NoIconHolder extends RecyclerView.ViewHolder {
        TextView txTitle;
        View contentRoot;

        NoIconHolder(View view) {
            super(view);
            contentRoot = view.findViewById(R.id.content_root);
            txTitle = view.findViewById(R.id.title);
        }
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/view/navigation/MenuListAdapter.kt
package dev.ragnarok.fenrir.view.navigation

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.card.MaterialCardView
import com.squareup.picasso3.Transformation
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.fragment.base.RecyclerBindableAdapter
import dev.ragnarok.fenrir.model.drawer.AbsMenuItem
import dev.ragnarok.fenrir.model.drawer.IconMenuItem
import dev.ragnarok.fenrir.model.drawer.NoIconMenuItem
import dev.ragnarok.fenrir.model.drawer.RecentChat
import dev.ragnarok.fenrir.picasso.PicassoInstance.Companion.with
import dev.ragnarok.fenrir.settings.CurrentTheme
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.Utils

class MenuListAdapter(
    context: Context,
    pageItems: MutableList<AbsMenuItem>,
    private val actionListener: ActionListener,
    private val paging: Boolean
) : RecyclerBindableAdapter<AbsMenuItem, RecyclerView.ViewHolder>(pageItems) {
    private val colorPrimary: Int = CurrentTheme.getColorPrimary(context)
    private val colorSurface: Int = CurrentTheme.getColorSurface(context)
    private val colorOnPrimary: Int = CurrentTheme.getColorOnPrimary(context)
    private val colorOnSurface: Int = CurrentTheme.getColorOnSurface(context)
    private val dp: Int = Utils.dpToPx(1f, context).toInt()
    private val transformation: Transformation = CurrentTheme.createTransformationForAvatar()
    private val noStroke: Boolean = Settings.get().other().is_side_no_stroke()
    override fun onBindItemViewHolder(
        viewHolder: RecyclerView.ViewHolder,
        position: Int,
        type: Int
    ) {
        val item = getItem(position)
        viewHolder.itemView.isSelected = item.isSelected
        when (type) {
            AbsMenuItem.TYPE_ICON -> if (paging || !noStroke) {
                bindIconHolder(viewHolder as NormalHolder, item as IconMenuItem)
            } else {
                bindIconNoStrokeHolder(viewHolder as NormalNoStrokeHolder, item as IconMenuItem)
            }
            AbsMenuItem.TYPE_RECENT_CHAT -> bindRecentChat(
                viewHolder as RecentChatHolder,
                item as RecentChat
            )
            AbsMenuItem.TYPE_WITHOUT_ICON -> bindWithoutIcon(
                viewHolder as NoIconHolder,
                item as NoIconMenuItem
            )
        }
    }

    private fun bindWithoutIcon(holder: NoIconHolder, item: NoIconMenuItem) {
        holder.txTitle.setText(item.title)
        holder.txTitle.setTextColor(if (item.isSelected) colorOnPrimary else colorOnSurface)
        holder.contentRoot.setOnClickListener { actionListener.onDrawerItemClick(item) }
        holder.contentRoot.setOnLongClickListener {
            actionListener.onDrawerItemLongClick(item)
            true
        }
    }

    private fun bindIconNoStrokeHolder(holder: NormalNoStrokeHolder, item: IconMenuItem) {
        holder.txtTitle.setText(item.title)
        holder.txtTitle.setTextColor(if (item.isSelected) colorOnPrimary else colorOnSurface)
        holder.tvCount.visibility = if (item.count > 0) View.VISIBLE else View.GONE
        holder.tvCount.text = item.count.toString()
        holder.tvCount.setTextColor(if (item.isSelected) colorOnPrimary else colorPrimary)
        holder.imgIcon.setImageResource(item.icon)
        holder.imgIcon.setColorFilter(if (item.isSelected) colorOnPrimary else colorOnSurface)
        holder.contentRoot.setBackgroundColor(if (item.isSelected) colorPrimary else colorSurface)
        holder.contentRoot.setOnClickListener { actionListener.onDrawerItemClick(item) }
        holder.contentRoot.setOnLongClickListener {
            actionListener.onDrawerItemLongClick(item)
            true
        }
    }

    private fun bindIconHolder(holder: NormalHolder, item: IconMenuItem) {
        holder.txtTitle.setText(item.title)
        holder.txtTitle.setTextColor(if (item.isSelected) colorOnPrimary else colorOnSurface)
        holder.tvCount.visibility = if (item.count > 0) View.VISIBLE else View.GONE
        holder.tvCount.text = item.count.toString()
        holder.tvCount.setTextColor(if (item.isSelected) colorOnPrimary else colorPrimary)
        holder.imgIcon.setImageResource(item.icon)
        holder.imgIcon.setColorFilter(if (item.isSelected) colorOnPrimary else colorOnSurface)
        holder.contentRoot.setCardBackgroundColor(if (item.isSelected) colorPrimary else colorSurface)
        holder.contentRoot.strokeWidth = if (item.isSelected) 0 else dp
        holder.contentRoot.setOnClickListener { actionListener.onDrawerItemClick(item) }
        holder.contentRoot.setOnLongClickListener {
            actionListener.onDrawerItemLongClick(item)
            true
        }
    }

    private fun bindRecentChat(holder: RecentChatHolder, item: RecentChat) {
        holder.tvChatTitle.text = item.title
        holder.tvChatTitle.setTextColor(if (item.isSelected) colorOnPrimary else colorOnSurface)
        if (item.iconUrl.isNullOrEmpty()) {
            with()
                .load(R.drawable.ic_group_chat)
                .transform(transformation)
                .into(holder.ivChatImage)
        } else {
            with()
                .load(item.iconUrl)
                .transform(transformation)
                .into(holder.ivChatImage)
        }
        (holder.contentRoot as MaterialCardView).setCardBackgroundColor(if (item.isSelected) colorPrimary else colorSurface)
        holder.contentRoot.setOnClickListener { actionListener.onDrawerItemClick(item) }
        holder.contentRoot.setOnLongClickListener {
            actionListener.onDrawerItemLongClick(item)
            true
        }
    }

    override fun viewHolder(view: View, type: Int): RecyclerView.ViewHolder {
        when (type) {
            AbsMenuItem.TYPE_DIVIDER -> return DividerHolder(view)
            AbsMenuItem.TYPE_RECENT_CHAT -> return RecentChatHolder(view)
            AbsMenuItem.TYPE_ICON -> return if (paging || !noStroke) NormalHolder(view) else NormalNoStrokeHolder(
                view
            )
            AbsMenuItem.TYPE_WITHOUT_ICON -> return NoIconHolder(view)
        }
        throw IllegalStateException()
    }

    override fun layoutId(type: Int): Int {
        when (type) {
            AbsMenuItem.TYPE_DIVIDER -> return if (noStroke) R.layout.drawer_list_item_divider_no_stroke else R.layout.drawer_list_item_divider
            AbsMenuItem.TYPE_RECENT_CHAT -> return R.layout.item_navigation_recents
            AbsMenuItem.TYPE_ICON -> return if (paging) R.layout.item_navigation else if (noStroke) R.layout.drawer_list_item_no_stroke else R.layout.drawer_list_item
            AbsMenuItem.TYPE_WITHOUT_ICON -> return R.layout.drawer_list_item_without_icon
        }
        throw IllegalStateException()
    }

    override fun getItemType(position: Int): Int {
        return getItem(position - headersCount).type
    }

    interface ActionListener {
        fun onDrawerItemClick(item: AbsMenuItem)
        fun onDrawerItemLongClick(item: AbsMenuItem)
    }

    private class NormalHolder(view: View) : RecyclerView.ViewHolder(view) {
        val imgIcon: ImageView = view.findViewById(R.id.icon)
        val txtTitle: TextView = view.findViewById(R.id.title)
        val tvCount: TextView = view.findViewById(R.id.counter)
        val contentRoot: MaterialCardView = view.findViewById(R.id.content_root)
    }

    private class NormalNoStrokeHolder(view: View) :
        RecyclerView.ViewHolder(view) {
        val imgIcon: ImageView = view.findViewById(R.id.icon)
        val txtTitle: TextView = view.findViewById(R.id.title)
        val tvCount: TextView = view.findViewById(R.id.counter)
        val contentRoot: ViewGroup = view.findViewById(R.id.content_root)
    }

    private class DividerHolder(itemView: View) : RecyclerView.ViewHolder(
        itemView
    )

    private class RecentChatHolder(itemView: View) :
        RecyclerView.ViewHolder(itemView) {
        val tvChatTitle: TextView = itemView.findViewById(R.id.title)
        val ivChatImage: ImageView = itemView.findViewById(R.id.avatar)
        val contentRoot: View = itemView.findViewById(R.id.content_root)
    }

    private class NoIconHolder(view: View) : RecyclerView.ViewHolder(view) {
        val txTitle: TextView = view.findViewById(R.id.title)
        val contentRoot: View = view.findViewById(R.id.content_root)
    }
}